desktop/createDownloadQueue.ts

```diff
@@ -0,0 +1,41 @@
+import type {BrowserWindow} from 'electron';
+import createQueue from '@libs/Queue/Queue';
+import ELECTRON_EVENTS from './ELECTRON_EVENTS';
+import electronDownload from './electronDownloadManager';
+import type {Options} from './electronDownloadManagerType';
+
+type DownloadItem = {
+    // The window where the download will be initiated
+    win: BrowserWindow;
+
+    // The URL of the file to be downloaded
+    url: string;
+
+    // The options for the download, such as save path, file name, etc.
+    options: Options;
+};
+
+const createDownloadQueue = () => {
+    const downloadItem = (item: DownloadItem): Promise<void> =>
+        new Promise((resolve) => {
+            const options = {
+                ...item.options,
+                onStarted: () => {
+                    item.win.webContents.send(ELECTRON_EVENTS.DOWNLOAD_STARTED, {url: item.url});
+                },
+                onCompleted: () => resolve(),
+                onCancel: () => resolve(),
```

Comment:
Why are these an empty promise?

Reply:
Since it uses a promise, we don't need to return any ID to the fileDownload function; just execute the resolve function passed.

Reply:
We simply need to perform the resolution, then download, whether completed or canceled, after which the queue will proceed with the next item. What do you think should be added here?

Reply:
I don't think that this is making a lot of sense right now. As I see it, there are the following issues with this approach:

1. Sending a response back to the window process from the `onStarted` handler would mean that a timeout will always occur (since you've set it to 5s) in the window process if there is something in the queue that is taking a lot of time to process. I think the event should be renamed as `DOWNLOAD_QUEUED` and should be sent back [here](https://github.com/Expensify/App/pull/37131/files#diff-e140a4e301565a464356efb0001afe4f8d8155cb3442473fed02f54fb5dd8065R616)

2. `electronDownload` function seems like a not very useful abstraction right now. I think the code would be much more understandable, if you'd just ditch the `options` interface and handle the downloading logic directly in this current function.

Reply:
Here's what I'm thinking on how this `downloadItem` function should be:
```
function queueItemProcessor(queueItem: DownloadItem) {
  return new Promise((resolve, reject) => {
     electronWindow.webContents.downloadURL(url);
     electronWindow.webContents.session.on('will-download', (item: DownloadItem) => {
       item.on('updated', () => {
         // Use the logic from electron-dl library and either `resolve` or `reject` the outermost promise.
       });
       item.on('complete', () => {
          // Use the logic from electron-dl library and either `resolve` or `reject` the outermost promise.
       });
     });
  });
}
```

Reply:
I believe the event should remain as DOWNLOAD_STARTED, and we should place `item.win.webContents.send(ELECTRON_EVENTS.DOWNLOAD_QUEUED, {url: item.url});` within the listener rather than `enqueueDownloadItem`. Changing it as suggested would cause the loading spinner to stop immediately upon clicking the download link.

Reply:
The loader should show till the download is either `completed` or `errored`. Said this, it might be even better to send to the browser process two more events (`DOWNLOAD_COMPLETED` and `DOWNLOAD_FAILED`).

Reply:
Sorry for the confusion.

The events that you had previously were fine. The issue with them was that if there is no response from the main process, the browser process would get indefinitely. We can easily make use of timeouts to get rid of this issue. Does that make sense?

Reply:
@allroundexperts what is the status for the event cancel? Is it failed/completed or new `DOWNLOAD_CANCELED`?

Please confirm that we no longer need the `DOWNLOAD_STARTED` event and it will be replaced by `DOWNLOAD_COMPLETED`, `DOWNLOAD_FAILED`, `DOWNLOAD_CANCELED`.

Reply:
I think all four of these should be there.

Reply:
`DOWNLOAD_STARTED` -> Called when the processor starts executing.
`DOWNLOAD_COMPLETED` -> When the file is downloaded
`DOWNLOAD_FAILED` -> When the file fails to download
`DOWNLOAD_CANCELLED` -> When the file download is cancelled

Reply:
@allroundexperts what should we do when `DOWNLOAD_STARTED`?


```
        window.electron.on(ELECTRON_EVENTS.DOWNLOAD_STARTED, (...args: unknown[]) => {
            const arg = Array.isArray(args) ? args[0] : null;
            const eventUrl = arg && typeof arg === 'object' && 'url' in arg ? arg.url : null;

            if (eventUrl === url) {
                clearTimeout(timeout);
                  // What should we do here? if we resolve here other events below is never reached 
            }
        });
    });


        window.electron.on(ELECTRON_EVENTS.DOWNLOAD_COMPLETED, (...args: unknown[]) => {
            const arg = Array.isArray(args) ? args[0] : null;
            const eventUrl = arg && typeof arg === 'object' && 'url' in arg ? arg.url : null;

            if (eventUrl === url) {
                clearTimeout(timeout);
                resolve();
            }
        });
    });

       window.electron.on(ELECTRON_EVENTS.DOWNLOAD_CANCELED, (...args: unknown[]) => {
            const arg = Array.isArray(args) ? args[0] : null;
            const eventUrl = arg && typeof arg === 'object' && 'url' in arg ? arg.url : null;

            if (eventUrl === url) {
                clearTimeout(timeout);
                resolve();
            }
        });
    });

       window.electron.on(ELECTRON_EVENTS.DOWNLOAD_FAILED, (...args: unknown[]) => {
            const arg = Array.isArray(args) ? args[0] : null;
            const eventUrl = arg && typeof arg === 'object' && 'url' in arg ? arg.url : null;

            if (eventUrl === url) {
                clearTimeout(timeout);
                resolve();
            }
        });
    });

```

Reply:
On the browser process, we don't need to listen for `DOWNLOAD_STARTED`. You can start the `loader` when the `DOWNLOAD_QUEUED` event is received.

Reply:
@allroundexperts I believe we do not need to start the loading spinner specifically when it's queued, as there is minimal delay between clicking on the file attachment and the item being queued. Currently, the loading starts immediately after the attachment file is clicked (`Download.setDownload`), which I think is acceptable.

https://github.com/Expensify/App/blob/807c945ef8edabe86cc62f16c866aa754db8750b/src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx#L46-L52

Reply:
Another thing to consider is -- start downloading spinner inside the queue requires us to create specific desktop code for the attachment onPress method.

Reply:
@wildan-m If we rely on `DOWNLOAD_STARTED`, then the loader won't start if there is already pending large download.

Reply:
@allroundexperts I fully agree with [your previous suggestion](https://github.com/Expensify/App/pull/37131#discussion_r1525358721). We trigger the `DOWNLOAD_STARTED` event to halt the loader, not to initiate it.

Reply:
@allroundexperts please elaborate further if I have misunderstood your query.